### ISSUE TITLE
RFE: explicit support for Verity-of-LUKS

### ISSUE THREAD

**Issue by michaelbeaumont** [systemd/systemd#33358]
### Component

systemd-gpt-auto-generator, systemd-repart

### Is your feature request related to a problem? Please describe

A number of systemd components either fail when encountering a dm-crypt partition on top of a dm-verity partition or do not support creating them in the first place.
I mostly have root filesystems in mind but this probably applies to `/usr` as well.

To be clear I mean the following:

```
# lsblk
sdb                    8:16   0    5G  0 disk
|-sdb1                 8:17   0    1G  0 part  /boot
|-sdb2                 8:18   0  100M  0 part
| `-root             254:0    0  2.9G  1 crypt
|   `-decrypted-root 254:1    0  2.9G  1 crypt /
|-sdb3                 8:19   0  2.9G  0 part
| `-root             254:0    0  2.9G  1 crypt
|   `-decrypted-root 254:1    0  2.9G  1 crypt /
# cryptsetup status /dev/mapper/root
/dev/mapper/root is active and is in use.
  type:    VERITY
# cryptsetup status /dev/mapper/decrypted-root
/dev/mapper/decrypted-root is active and is in use.
  type:    LUKS2
  cipher:  aes-xts-plain64
  keysize: 512 bits
  key location: keyring
  device:  /dev/mapper/root
  sector size:  512
  offset:  32768 sectors
  size:    6062608 sectors
  mode:    readonly
```

This is maybe [mentioned in the TODO](https://github.com/systemd/systemd/blob/2a00e925988435cfde5ee38c1ab47a1ed07ef37f/TODO#L344) but it's not clear if this is dm-verity on dm-crypt or dm-crypt on dm-verity

### Describe the solution you'd like

At least:
* `systemd-gpt-auto-generator` should be updated to support automatic opening of a LUKS volume after opening verity partitions
* `systemd-repart` should support `Encrypt=key-file` and `Verity=data` as well as `CopyBlocks=auto`, there may be more relevant options
* `systemd-sysupdate` with `Path=auto`

Is there interest in supporting this use case?

I've made a couple changes to make some things work:

* updating `block_device_get_originating` to search "one more level up", this fixes `systemd-repart` and `systemd-sysupdate`
* just removing the check for `Encrypt=key-file` and `Verity=data` together "just works", except that `Minimize=` on the hash partition no longer works. I think this is due to ordering of encrypting and minimizing
* one thing that doesn't work yet is `systemd-repart` and `CopyBlocks=auto` finding the right partitions, I think `resolve_copy_blocks_auto` needs to be updated here.

### The systemd version you checked that didn't have the feature you are asking for

255/main

**Comment by poettering** [comment]
yeah, we should have explicit support for something like that. But I'd assume we'd want dm-verity-of-dm-crypt rather than the opposite. (simply because we store sigs of verity stuff in separate partitions, and choose partition uuids based on the root hash, and none of that would make sense anymore if we'd do it te other way round).

what's your rationale for wanting dm-crypt-of-dm-verity?

**Comment by michaelbeaumont** [comment]
We might mean the same thing? I want dm-verity to integrity check an encrypted data partition. I meant "on top" in the sense that we run veritysetup first and then cryptsetup.

Feel free to change the issue name/description to make it as clear as possible!

My particular use case is to use `roothash` to decide which encrypted partition to use as root file system and unlock.

**Comment by Plan-V64** [comment]
Just a note of support for the proposed layering:

I believe that the other way around (i.e. the dm-verity layer on top of the dm-crypt layer) (i.e. run cryptsetup first and then veritysetup) would fit the Encrypt-and-MAC (anti-)pattern. Because dm-verity stores the hashes on a separate partition, not inside its main "container" partition.

The Encrypt-and-MAC construction tends to be more vulnerable to cryptographic attack than the Encrypt-then-MAC construction. For example, the attacker can recognize known/repeated cleartext blocks. (That particular vulnerability can be "patched up" by adding a block sequence number to the MAC/hash. But solving such issues is not a design goal of dm-verity.)

So the @michaelbeaumont ordering is fine, the other way around would not be a good option.

Disclaimer: I am not at all an experienced cryptographer; get advice from one.
